lib/hsa/mcwamp_hsa.cpp

```diff
@@ -4890,6 +4894,7 @@ inline void
 HSABarrier::wait() {
     // Wait on completion signal until the barrier is finished
     hsa_signal_wait_scacquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+    dispose();
```

Comment:
Calling `dispose()` in `wait()` doesn't seem right... 
If `wait()` is not called before the `dtor()` of HSAOp, will it cause a memory leak?
Would it be better to call `dispose()` from the `dtor()`? Then, you probably won't need the extra atomic flag since it's supposed to be destroyed by the last owning thread?

Reply:
Admittedly, this is the least straightforward part of this PR.  Yes, `wait()` must be called before calling `dispose()`.  This is done in all `dtor()`s.  The reason for calling `dispose()` inside the `wait()` is to decrement the ref counts of stored dep ops as soon as possible.  This is what we were doing before this PR, but this PR accidentally changed that behavior to only decrement dep ops in the `dtor()`.  `dispose()` will free everything but the `_signal` --- the final `_signal` cleanup was moved to the `HSAOp` `dtor`.  `dispose()` should only be called by a single thread, that's why the `atomic_flag` was added.  You're right that the `dtor()` will only be destroyed by the last calling thread, and it will `wait()` just in case.  But any thread can call `wait()` and we should clean up resources there greedily.